File: 307. Range Sum Query - Mutable/range-sum-query-mutable.py
class NumArray:
    class SegmentTree:
        def __init__(self, leftmost, rightmost, a):
            self.leftmost = leftmost
            self.rightmost = rightmost
            if leftmost == rightmost:
                self.sum = a[leftmost]
            else:
                mid = (leftmost + rightmost) // 2
                self.lChild = self.__class__(leftmost, mid, a)
                self.rChild = self.__class__(mid + 1, rightmost, a)
                self.recalc()
        def recalc(self):
            if self.leftmost == self.rightmost: return
            self.sum = self.lChild.sum + self.rChild.sum
        def pointUpdate(self, index, newVal):
            if self.leftmost == self.rightmost:
                self.sum = newVal
                return
            if index <= self.lChild.rightmost:
                self.lChild.pointUpdate(index, newVal)
            else:
                self.rChild.pointUpdate(index, newVal)
            self.recalc()
        def rangeSum(self, l, r):
            if l > self.rightmost or r < self.leftmost:
                return 0
            elif l <= self.leftmost and r >= self.rightmost:
                return self.sum
            return self.lChild.rangeSum(l, r) + self.rChild.rangeSum(l, r) 


    def __init__(self, nums: List[int]):
        # Build: O(n)
        self.st = self.SegmentTree(0, len(nums) - 1, nums)
        

    def update(self, index: int, val: int) -> None:
        # O(log(n))
        self.st.pointUpdate(index, val)

    def sumRange(self, left: int, right: int) -> int:
        # O(log(n))
        return self.st.rangeSum(left, right)
        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: 307. Range Sum Query - Mutable/range-sum-query-mutable.java
class NumArray {
    // The number of nodes is 2n - 1
    // n + n/2 + n/4 + ... + 1 
    //   = \sum_{i=0}^{k} n(1/2)^i   -> last term: n (1/2)^k = 1
    //   = \sum_{i=0}^{\log{n}} n(1/2)^i 
    //   = n (1 - 0.5^{\log{n} + 1}) / (1 - 0.5) = 2n (1 - 1/2n)
    //   = 2n - 1
    class SegmentTree {
        SegmentTree lChild, rChild;
        int leftmost, rightmost;
        int sum;
        SegmentTree(int leftmost, int rightmost, int[] a) {
            this.leftmost = leftmost;
            this.rightmost = rightmost;
            if (leftmost == rightmost) {
                this.sum = a[leftmost];
            } else {
                int mid = leftmost + (rightmost - leftmost) / 2;
                lChild = new SegmentTree(leftmost, mid, a);
                rChild = new SegmentTree(mid + 1, rightmost, a);
                recalc();
            }
        }
        void recalc(){
            if (leftmost == rightmost) return;
            this.sum = lChild.sum + rChild.sum;
        }
        int rangeSum(int l, int r) {
            if (l <= leftmost && r >= rightmost) {
                return sum;
            } else if (r < leftmost || l > rightmost ) {
                return 0;
            }
            return lChild.rangeSum(l, r) + rChild.rangeSum(l, r);
        }
        void pointUpdate(int index, int val) {
            if (leftmost == rightmost) {
                sum = val;
                return;
            }
            if (index <= lChild.rightmost) {
                lChild.pointUpdate(index, val);
            } else rChild.pointUpdate(index, val);
            recalc();
        }
    }
    private SegmentTree st;
    public NumArray(int[] nums) {
        st = new SegmentTree(0, nums.length - 1, nums);
    }
    
    public void update(int index, int val) {
        st.pointUpdate(index, val);
    }
    
    public int sumRange(int left, int right) {
        return st.rangeSum(left, right);
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(index,val);
 * int param_2 = obj.sumRange(left,right);
 */
